README.md
# Redis Sentinel Exporter for Prometheus

This is a simple server that scrapes Redis Sentinel stats and exports them via HTTP for Prometheus consumption.

## Configuration

* `-debug` (env `DEBUG`) – Output verbose debug information.
* `-log-format` (env `LOG_FORMAT`) – Log format, valid options are txt and json. (default `txt`)
* `-namespace` (env `NAMESPACE`) – Namespace for metrics. (default `redis_sentinel`)
* `-sentinel.addr` (env `SENTINEL_ADDR`) – Redis Sentinel host:port. (default `redis://127.0.0.1:26379`)
* `-sentinel.password` (env `SENTINEL_PASSWORD`) – Redis Sentinel password (optional).
* `-sentinel.password-file` (env `SENTINEL_PASSWORD_FILE`) - Path to Redis Sentinel password file (optional).
* `-version` – Prints version and exit.
* `-web.listen-address` (env `LISTEN_ADDRESS`) – Address to listen on for web interface and telemetry. (default `:9355`)
* `-web.telemetry-path` (env `TELEMETRY_PATH`) – Path under which to expose metrics. (default `/metrics`)

## Links

* [Binary](https://github.com/leominov/redis_sentinel_exporter/releases)
* [Docker Image](https://hub.docker.com/r/leominov/redis_sentinel_exporter)
* [Grafana Dashboard](https://grafana.com/dashboards/9570)

## Forked from

* <https://github.com/hiboxsystems/redis_sentinel_exporter>
* <https://github.com/leominov/redis_sentinel_exporter>

## Docker

Images are pushed to <https://hub.docker.com/repository/docker/monotek/redis-sentinel-exporter>

Run via:

```bash
docker run -it --rm -p 9355:9355 monotek/redis-sentinel-exporter
```


CHANGELOG.md
# Changelog

## 1.8.0 / 2022-05-24

- updates go dependencies
- change docker entrypoint

## 1.7.2 / 2022-03-01

- [FEATURE] First upload by temporary maintainer.


helpers_test.go
package main

import (
	"os"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestGetStringEnv(t *testing.T) {
	os.Unsetenv("FOOBAR")
	assert.Equal(t, "default", GetStringEnv("FOOBAR", "default"))
	os.Setenv("FOOBAR", "specified")
	assert.Equal(t, "specified", GetStringEnv("FOOBAR", "default"))
}

func TestGetBoolEnv(t *testing.T) {
	os.Unsetenv("FOOBAR")
	assert.Equal(t, true, GetBoolEnv("FOOBAR", true))
	os.Setenv("FOOBAR", "false")
	assert.Equal(t, false, GetBoolEnv("FOOBAR", true))
	os.Setenv("FOOBAR", "1")
	assert.Equal(t, true, GetBoolEnv("FOOBAR", false))
	os.Setenv("FOOBAR", "ABCD")
	assert.Equal(t, false, GetBoolEnv("FOOBAR", false))
}


exporter_test.go
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestNewRedisSentinelExporter(t *testing.T) {
	e := NewRedisSentinelExporter("127.0.0.1:8080", "ns", "")
	assert.NotNil(t, e)
}


sentinel_info_test.go
package main

import (
	"bytes"
	"io/ioutil"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestParseValue(t *testing.T) {
	cases := map[string]interface{}{
		"ok":     1.0,
		"down":   0.0,
		"fail":   0.0,
		"13.0":   13.0,
		"foobar": "foobar",
	}
	for in, out := range cases {
		assert.Equal(t, out, ParseValue(in))
	}
}

func TestParseMasterInfo(t *testing.T) {
	masterA := ParseMasterInfo("foobar,foo=bar,name=mymaster,status=ok,address=127.0.0.1:6379,slaves=2,sentinels=3")
	assert.NotNil(t, masterA)
	assert.Equal(t, "mymaster", masterA.Metrics["name"].(string))

	_, ok := masterA.Metrics["foobar"]
	assert.False(t, ok)

	_, ok = masterA.Metrics["foo"]
	assert.False(t, ok)
	assert.Equal(t, 1.0, masterA.Metrics["status"].(float64))
}

func TestParseInfo(t *testing.T) {
	tests := []struct {
		filename string
		master   bool
	}{
		{
			filename: "test_data/case-1",
			master:   true,
		},
		{
			filename: "test_data/case-2",
		},
	}
	for _, test := range tests {
		b, err := ioutil.ReadFile(test.filename)
		if err != nil {
			t.Fatal(err)
		}
		b = bytes.Replace(b, []byte("\n"), []byte("\r\n"), -1)
		metricRequiredKeys := metricBuildInfo
		for metricName := range metricMap {
			metricRequiredKeys = append(metricRequiredKeys, metricName)
		}
		si := ParseInfo(string(b), metricRequiredKeys, test.master)
		assert.NotNil(t, si)
	}
}


e2e.sh
#!/usr/bin/env bash
# Usage:   ./e2e.sh ${RedisVersion} ${FixtureVersion} ${PasswordProtected} ${PasswordFromFile}
# Example: ./e2e.sh 6.0.6 5.0 1 0

set -e
set -x
set -o pipefail
set -o nounset

redis_version="$1"
fixture_version="$2"
require_pass="$3"
pass_from_file="$4"

skip_re="^(redis_sentinel_build_info|redis_sentinel_exporter_build_info|redis_sentinel_info|redis_sentinel_used_cpu|redis_sentinel_exporter_last_scrape_duration_seconds|redis_sentinel_uptime_in_seconds|redis_sentinel_connections_received_total|redis_sentinel_net|redis_sentinel_instantaneous|redis_sentinel_process_id)"

echo "==> Redis $redis_version"

rm -rf "redis-${redis_version}" "redis-${redis_version}.tar.gz"

wget "http://download.redis.io/releases/redis-${redis_version}.tar.gz"
tar -zxvf "redis-${redis_version}.tar.gz"
cd "redis-${redis_version}" || exit
make

nohup ./src/redis-server &

success="0"
for i in {1..5}; do
  echo "Run test nr: $i"
  if ./src/redis-cli -p 6379 PING; then
    success="1"
    break
  fi
  sleep 1
done

if [[ $success == "0" ]]; then
  echo "Redis PING failed"
  exit 1
fi

if [[ $require_pass == "0" ]]; then
  cp ../test_data/sentinel.conf sentinel.conf
else
  cp ../test_data/sentinel-protected.conf sentinel.conf
fi

cat sentinel.conf
nohup ./src/redis-sentinel sentinel.conf &

success="0"
for i in {1..5}; do
  echo "Run test nr: $i"
  if ./src/redis-cli -p 26379 PING; then
    success="1"
    break
  fi
  sleep 1
done

if [[ $success == "0" ]]; then
  echo "Redis Sentinel PING failed"
  exit 1
fi

cd ../

go build

if [[ $require_pass == "0" ]]; then
  nohup ./redis_sentinel_exporter --debug &
else
  if [[ $pass_from_file == "0" ]]; then
    nohup ./redis_sentinel_exporter --debug --sentinel.password=ABCD &
  else
    echo "ABCD" > /tmp/sentinel_password.txt
    nohup ./redis_sentinel_exporter --debug --sentinel.password-file=/tmp/sentinel_password.txt &
  fi
fi

wget --retry-connrefused --tries=5 -O - "127.0.0.1:9355/metrics"| grep "redis_" | grep -E -v "${skip_re}" > "e2e-output.txt"

diff -u \
  "test_data/e2e-output-v${fixture_version}.txt" \
  "e2e-output.txt"


exporter.go
package main

import (
	"strings"
	"time"

	"github.com/gomodule/redigo/redis"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/sirupsen/logrus"
)

var (
	metricBuildInfo = []string{
		"redis_version",
		"redis_build_id",
		"redis_mode",
	}
	metricMap = map[string]string{
		// Server
		"uptime_in_seconds": "uptime_in_seconds",
		"process_id":        "process_id",
		// Clients
		"connected_clients":          "connected_clients",
		"client_longest_output_list": "client_longest_output_list",
		"client_biggest_input_buf":   "client_biggest_input_buf",
		"blocked_clients":            "blocked_clients",
		// Stats
		"total_connections_received": "connections_received_total",
		"total_commands_processed":   "commands_processed_total",
		"instantaneous_ops_per_sec":  "instantaneous_ops_per_sec",
		"total_net_input_bytes":      "net_input_bytes_total",
		"total_net_output_bytes":     "net_output_bytes_total",
		"instantaneous_input_kbps":   "instantaneous_input_kbps",
		"instantaneous_output_kbps":  "instantaneous_output_kbps",
		"rejected_connections":       "rejected_connections_total",
		"expired_keys":               "expired_keys_total",
		"evicted_keys":               "evicted_keys_total",
		"keyspace_hits":              "keyspace_hits_total",
		"keyspace_misses":            "keyspace_misses_total",
		"pubsub_channels":            "pubsub_channels",
		"pubsub_patterns":            "pubsub_patterns",
		"latest_fork_usec":           "latest_fork_usec",
		// CPU
		"used_cpu_sys":           "used_cpu_sys",
		"used_cpu_user":          "used_cpu_user",
		"used_cpu_sys_children":  "used_cpu_sys_children",
		"used_cpu_user_children": "used_cpu_user_children",
		// Sentinel
		"sentinel_masters":                "masters",
		"sentinel_tilt":                   "tilt",
		"sentinel_running_scripts":        "running_scripts",
		"sentinel_scripts_queue_length":   "scripts_queue_length",
		"sentinel_simulate_failure_flags": "simulate_failure_flags",
	}
)

type Exporter struct {
	addr             string
	namespace        string
	sentinelPassword string
	metrics          map[string]*prometheus.GaugeVec
	duration         prometheus.Gauge
	scrapeErrors     prometheus.Gauge
	totalScrapes     prometheus.Counter
}

func NewRedisSentinelExporter(addr, namespace string, sentinelPassword string) *Exporter {
	e := &Exporter{
		addr:             addr,
		namespace:        namespace,
		sentinelPassword: sentinelPassword,
		duration: prometheus.NewGauge(prometheus.GaugeOpts{
			Namespace: namespace,
			Name:      "exporter_last_scrape_duration_seconds",
			Help:      "The last scrape duration.",
		}),
		totalScrapes: prometheus.NewCounter(prometheus.CounterOpts{
			Namespace: namespace,
			Name:      "exporter_scrapes_total",
			Help:      "Current total redis scrapes.",
		}),
		scrapeErrors: prometheus.NewGauge(prometheus.GaugeOpts{
			Namespace: namespace,
			Name:      "exporter_last_scrape_error",
			Help:      "The last scrape error status.",
		}),
	}
	e.initGauges()
	return e
}

func (e *Exporter) initGauges() {
	e.metrics = map[string]*prometheus.GaugeVec{}

	e.metrics["info"] = prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Namespace: e.namespace,
		Name:      "info",
		Help:      "Information about Sentinel",
	}, []string{"version", "build_id", "mode"})

	// Masters
	e.metrics["master_status"] = prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Namespace: e.namespace,
		Name:      "master_status",
		Help:      "Status of master",
	}, []string{"name", "address"})
	e.metrics["master_slaves"] = prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Namespace: e.namespace,
		Name:      "master_slaves",
		Help:      "Slaves of master",
	}, []string{"name", "address"})
	e.metrics["master_sentinels"] = prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Namespace: e.namespace,
		Name:      "master_sentinels",
		Help:      "Sentinels of master",
	}, []string{"name", "address"})

	// All other metrics
	for metricOrigName, metricOutName := range metricMap {
		e.metrics[metricOrigName] = prometheus.NewGaugeVec(prometheus.GaugeOpts{
			Namespace: e.namespace,
			Name:      metricOutName,
			Help:      metricOutName,
		}, []string{})
	}
}

func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
	ch <- e.duration.Desc()
	ch <- e.totalScrapes.Desc()
	ch <- e.scrapeErrors.Desc()

	for _, m := range e.metrics {
		m.Describe(ch)
	}
}

func (e *Exporter) scrapeInfo() (string, error) {
	options := []redis.DialOption{
		redis.DialConnectTimeout(5 * time.Second),
		redis.DialReadTimeout(5 * time.Second),
		redis.DialWriteTimeout(5 * time.Second),
		redis.DialPassword(e.sentinelPassword),
	}

	logrus.Debugf("Trying DialURL(): %s", e.addr)
	c, err := redis.DialURL(e.addr, options...)

	if err != nil {
		logrus.Debugf("DialURL() failed, err: %s", err)
		if frags := strings.Split(e.addr, "://"); len(frags) == 2 {
			logrus.Debugf("Trying: Dial(): %s %s", frags[0], frags[1])
			c, err = redis.Dial(frags[0], frags[1], options...)
		} else {
			logrus.Debugf("Trying: Dial(): tcp %s", e.addr)
			c, err = redis.Dial("tcp", e.addr, options...)
		}
	}

	if err != nil {
		logrus.Debugf("aborting for addr: %s - redis sentinel err: %s", e.addr, err)
		return "", err
	}

	defer c.Close()
	logrus.Debugf("connected to: %s", e.addr)

	body, err := redis.String(c.Do("info"))
	if err != nil {
		logrus.Debugf("cannot execute command info: %v", err)
		return "", err
	}

	return body, nil
}

func (e *Exporter) setMetrics(i *SentinelInfo) {
	for metricName, gauge := range e.metrics {
		switch metricName {
		case "info":
			gauge.WithLabelValues(
				i.Metrics["redis_version"].(string),
				i.Metrics["redis_build_id"].(string),
				i.Metrics["redis_mode"].(string),
			).Set(float64(1))
		case "master_status", "master_slaves", "master_sentinels":
			gauge.Reset()
			metricType := strings.TrimPrefix(metricName, "master_")
			for _, m := range i.Masters {
				gauge.WithLabelValues(
					m.Metrics["name"].(string),
					m.Metrics["address"].(string),
				).Set(m.Metrics[metricType].(float64))
			}
		default:
			if _, ok := i.Metrics[metricName]; ok {
				gauge.WithLabelValues().Set(i.Metrics[metricName].(float64))
			}
		}
	}
}

func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
	var errorCount int
	now := time.Now().UnixNano()
	e.totalScrapes.Inc()

	ch <- e.duration
	ch <- e.totalScrapes
	ch <- e.scrapeErrors

	infoRaw, err := e.scrapeInfo()
	metricRequiredKeys := metricBuildInfo
	for metricName := range metricMap {
		metricRequiredKeys = append(metricRequiredKeys, metricName)
	}
	if err != nil {
		errorCount++
	} else {
		sentinelInfo := ParseInfo(infoRaw, metricRequiredKeys, true)
		e.setMetrics(sentinelInfo)
	}

	e.scrapeErrors.Set(float64(errorCount))
	e.duration.Set(float64(time.Now().UnixNano()-now) / 1000000000)

	for _, m := range e.metrics {
		m.Collect(ch)
	}
}


sentinel_info.go
package main

import (
	"strconv"
	"strings"

	"github.com/sirupsen/logrus"
)

var (
	masterMetricMap = map[string]string{
		"name":      "name",
		"status":    "status",
		"address":   "address",
		"slaves":    "slaves",
		"sentinels": "sentinels",
	}
)

type SentinelInfo struct {
	Metrics map[string]interface{}
	Masters []*Master
}

type Master struct {
	Metrics map[string]interface{}
}

// Format:
// name=mymaster,status=ok,address=172.17.8.101:6379,slaves=2,sentinels=3
func ParseMasterInfo(info string) *Master {
	split := strings.Split(info, ",")
	m := &Master{
		Metrics: make(map[string]interface{}),
	}
	for _, keyPair := range split {
		s := strings.Split(keyPair, "=")
		if len(s) != 2 {
			continue
		}
		fieldKey := s[0]
		fieldValue := s[1]
		for metricOriginalName, metricName := range masterMetricMap {
			if metricOriginalName != fieldKey {
				continue
			}
			m.Metrics[metricName] = ParseValue(fieldValue)
		}
	}
	return m
}

func ParseValue(value string) interface{} {
	if value == "ok" {
		return float64(1)
	} else if value == "fail" || strings.Contains(value, "down") {
		return float64(0)
	} else if val, err := strconv.ParseFloat(value, 64); err == nil {
		return val
	}
	return value
}

func ParseInfo(info string, keys []string, includeMasters bool) *SentinelInfo {
	lines := strings.Split(info, "\r\n")
	i := &SentinelInfo{
		Metrics: make(map[string]interface{}),
	}
	for _, line := range lines {
		if len(line) == 0 {
			continue
		}
		if strings.HasPrefix(line, "#") {
			continue
		}
		logrus.Debugf("info: %s", line)
		split := strings.SplitN(line, ":", 2)
		if len(split) < 2 {
			continue
		}
		fieldKey := split[0]
		fieldValue := split[1]
		if strings.HasPrefix(fieldKey, "master") && includeMasters {
			master := ParseMasterInfo(fieldValue)
			i.Masters = append(i.Masters, master)
			continue
		}
		for _, key := range keys {
			if key != fieldKey {
				continue
			}
			i.Metrics[key] = ParseValue(fieldValue)
		}
	}
	return i
}


main.go
package main

import (
	"flag"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"strings"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"github.com/prometheus/common/version"
	"github.com/sirupsen/logrus"
)

var (
	isDebug              = flag.Bool("debug", GetBoolEnv("DEBUG", false), "Output verbose debug information.")
	listenAddress        = flag.String("web.listen-address", GetStringEnv("LISTEN_ADDRESS", ":9355"), "Address to listen on for web interface and telemetry.")
	logFormat            = flag.String("log-format", GetStringEnv("LOG_FORMAT", "txt"), "Log format, valid options are txt and json.")
	metricPath           = flag.String("web.telemetry-path", GetStringEnv("TELEMETRY_PATH", "/metrics"), "Path under which to expose metrics.")
	namespace            = flag.String("namespace", GetStringEnv("NAMESPACE", "redis_sentinel"), "Namespace for metrics.")
	sentinelAddr         = flag.String("sentinel.addr", GetStringEnv("SENTINEL_ADDR", "redis://127.0.0.1:26379"), "Redis Sentinel host:port.")
	sentinelPassword     = flag.String("sentinel.password", GetStringEnv("SENTINEL_PASSWORD", ""), "Redis Sentinel password (optional).")
	sentinelPasswordFile = flag.String("sentinel.password-file", GetStringEnv("SENTINEL_PASSWORD_FILE", ""), "Path to Redis Sentinel password file (optional).")
	versionPrint         = flag.Bool("version", false, "Prints version and exit.")
)

func main() {
	flag.Parse()

	if *versionPrint {
		fmt.Println(version.Print("redis sentinel exporter"))
		os.Exit(0)
	}

	switch *logFormat {
	case "json":
		logrus.SetFormatter(&logrus.JSONFormatter{})
	default:
		logrus.SetFormatter(&logrus.TextFormatter{})
	}

	logrus.Infof("Starting Redis Sentinel Exporter %s...", version.Version)

	if *isDebug {
		logrus.SetLevel(logrus.DebugLevel)
		logrus.Debug("Enabling debug output")
	}

	if len(*sentinelAddr) == 0 {
		logrus.Fatal("Must specify a non-empty sentinel.addr")
	}

	password := *sentinelPassword
	if len(*sentinelPasswordFile) > 0 {
		body, err := ioutil.ReadFile(*sentinelPasswordFile)
		if err != nil {
			logrus.WithError(err).Fatal("Failed to load Redis Sentinel password file")
		}
		password = strings.TrimSpace(string(body))
	}

	exporter := NewRedisSentinelExporter(*sentinelAddr, *namespace, password)

	prometheus.MustRegister(exporter)

	// Deprecated and will be removed in a future version,
	// use `redis_sentinel_exporter_build_info`
	prometheus.MustRegister(version.NewCollector(*namespace))
	versionNamespace := *namespace + "_exporter"
	prometheus.MustRegister(version.NewCollector(versionNamespace))

	http.Handle(*metricPath, promhttp.Handler())

	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte(`
			<html>
			<head><title>Redis Sentinel Exporter ` + version.Version + `</title></head>
			<body>
			<h1>Redis Sentinel Exporter ` + version.Version + `</h1>
			<p><a href='` + *metricPath + `'>Metrics</a></p>
			</body>
			</html>
		`))
	})

	logrus.Printf("Providing metrics at %s%s", *listenAddress, *metricPath)
	logrus.Fatal(http.ListenAndServe(*listenAddress, nil))
}
